Ensure endpoint validation occurs before initial regeneration

tgraf opened the pull request:
Due to a performance optimization, validateEndpoint() was changed to use
the pod store instad of contacting the apiserver directly. This required
to make the validation async to wait until the pod cache is available.

This change however caused the endpoint to be exposed and regenerated
before the endpoint was fully validated. This in turn caused the
validation to occur while the regeneration was being performed. When the
validation then failed, the endpoint was deleted while the endpoint was
being regenerated, causing the endpoint to fail with a "JoinEP" message
in the logs.

This commit requires the endpoint to be validated before exposure and
initial regeneration.

This problem only exists in master. No need to backport.

Fixes: #11645
Fixes: #11076
Fixes: #10856
Fixes: #10667
Fixes: #10278
Fixes: 9975bba1637 ("pkg/endpoint: fetch pod and namespace labels from local stores")

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -395,6 +387,20 @@ func NewDaemon(ctx context.Context, dp datapath.Datapath) (*Daemon, *endpointRes
 	treatRemoteNodeAsHost := option.Config.AlwaysAllowLocalhost() && !option.Config.EnableRemoteNodeIdentity
 	policyApi.InitEntities(option.Config.ClusterName, treatRemoteNodeAsHost)
 
+	d.k8sCachesSynced = d.k8sWatcher.InitK8sSubsystem()
```
Not sure if I'm missing something here, but shouldn't we guard this statement by checking if K8s is enabled? What would happen if K8s is disabled?

tgraf replied:
It was not protected before which is why I've kept it this way. The check is embedded in `EnableK8sWatcher()`.

christarazi reviewed (commented):
The code looks good to me, besides one question that I have.

tgraf committed:
k8s: Export K8sAPIGroupPodV1Core

A following commit will require to wait on the K8sAPIGroupPodV1Core
cache

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
endpoint: Ensure restored endpoint is validated before regenerating it

Due to a performance optimization, validateEndpoint() was changed to use
the pod store instad of contacting the apiserver directly. This required
to make the validation async to wait until the pod cache is available.

This change however caused the endpoint to be exposed and regenerated
before the endpoint was fully validated. This in turn caused the
validation to occur while the regeneration was being performed. When the
validation then failed, the endpoint was deleted while the endpoint was
being regenerated, causing the endpoint to fail with a "JoinEP" message
in the logs.

This commit requires the endpoint to be validated before exposure and
initial regeneration.

Fixes: #11645
Fixes: #11076
Fixes: #10856
Fixes: #10667
Fixes: #10278
Fixes: 9975bba1637 ("pkg/endpoint: fetch pod and namespace labels from local stores")

Signed-off-by: Thomas Graf <thomas@cilium.io>
